sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3927,8 +3927,12 @@ object TimeWindowing extends Rule[LogicalPlan] {
           val projections = windows.map(_ +: child.output)
 
           val filterExpr =
-            window.timeColumn >= windowAttr.getField(WINDOW_START) &&
-              window.timeColumn < windowAttr.getField(WINDOW_END)
+            if (window.windowDuration % window.slideDuration == 0) {
```

Comment:
Probably good to leave code comment like below:

> When the condition `windowDuration % slideDuration = 0` is fulfilled, the estimation of the number of windows becomes exact one, which means all produced windows are valid.

I'm not a native so the sentence may not be perfect, but may be acceptable and understandable.

Reply:
@HeartSaVioR  thanks!

I've added code comments as you suggested. I think your suggestion is good enough.